Add validation of agent flag values for ConfigMap

romanspb80 opened the pull request:
If the Cilium agent flags are passed via a mounted ConfigMap
(cilium-agent --config-dir=/tmp/cilium/config-map), the default for Helm
deployments, the flag values are not validated. For example if you set
"restore" with invalid value "0SO##ME5_RANDOM" in ConfigMap then Agent
would run with incorrect parameter:
.....
level=info msg=" --restore='0SO##ME5_RANDOM'" subsys=daemon
.....

But if start Agent with CLI then the validation will warn and prevent
starting the agent:
cilium-agent[8654]: invalid argument "0SO##ME5_RANDOM" for "--restore"
flag: strconv.ParseBool: parsing "0SO##ME5_RANDOM": invalid syntax

This commit add agent flag values validation for ConfigMap

Fixes: #13070

Signed-off-by: Roman Ptitcyn romanspb@yahoo.com

anfernee commented on pkg/option/config.go:
```diff
@@ -3113,8 +3114,23 @@ func InitConfig(programName, configName string) func() {
+					if val := fmt.Sprintf("%v", value); val != "" {
+						flags := cmd.Flags()
+						// check whether the flag exists
+						if flag := flags.Lookup(key); flag != nil {
```
If  `flag` is not found, it should be an error as well, right?

romanspb80 replied:
> If `flag` is not found, it should be an error as well, right?

No. The validation works for only existing flags. In case of unknown flags Cilium will start with these unknown flags.

anfernee commented on pkg/option/config.go:
```diff
@@ -3113,8 +3114,23 @@ func InitConfig(programName, configName string) func() {
-				if err != nil {
+
+				// validate the config-map
+				var msgError = "Incorrect config-map flag"
```
I don't think a variable is needed here.

romanspb80 replied:
> I don't think a variable is needed here.

Why not. Suggest another option

pchaigno replied:
I agree with Yongkun, just passing that text to `Fatal` is enough.

romanspb80 replied:
Sorry. I misunderstood the remark... :) . Yes. It's better this way.
Done

anfernee commented on pkg/option/config.go:
```diff
@@ -3113,8 +3114,23 @@ func InitConfig(programName, configName string) func() {
+						if flag := flags.Lookup(key); flag != nil {
+							// validate the value of flag
+							if err := flag.Value.Set(val); err != nil {
+								log.WithError(err).Fatal(msgError + "  " + key)
```
I think we don't need to append `key` in fatal message. I assume `err` will contain the useful information.

romanspb80 replied:
> I think we don't need to append `key` in fatal message. I assume `err` will contain the useful information.

No. 'err' doesn't contain information regarding 'key'. It shows info about value only:
error="strconv.ParseBool: parsing \"0SO##ME5_RANDOM\": invalid syntax"

errordeveloper commented on pkg/option/config.go:
```diff
@@ -3113,8 +3114,23 @@ func InitConfig(programName, configName string) func() {
+							}
+						}
+					}
+				}
```
This should a separate function, the nesting is way too deep. I understand this package has too many lines already, but let's not add deeply nested logic here.

romanspb80 replied:
> This should a separate function, the nesting is way too deep. I understand this package has too many lines already, but let's not add deeply nested logic here.

Done

errordeveloper reviewed (commented):
I think overall it's a good idea to add validation, however we need to discuss whether it's expected that unknown flags are ignored in case of an upgrade or downgrade.

errordeveloper commented on pkg/option/config.go:
```diff
@@ -3113,8 +3114,23 @@ func InitConfig(programName, configName string) func() {
+						if flag := flags.Lookup(key); flag != nil {
+							// validate the value of flag
+							if err := flag.Value.Set(val); err != nil {
+								log.WithError(err).Fatal(msgError + "  " + key)
```
Perhaps this just needs to be a warning?

romanspb80 replied:
> Perhaps this just needs to be a warning?

No. The ConfigMap validation should be same as for CLI.

jibi commented on pkg/option/config.go:
```diff
@@ -3164,8 +3165,25 @@ func sanitizeIntParam(paramName string, paramDefault int) int {
+			if flag := flags.Lookup(key); flag != nil {
+				// validate the value of flag
+				if err := flag.Value.Set(val); err != nil {
+					log.WithError(err).Fatal("Incorrect config-map flag " + key)
```
nit: I'd change this function to return an error and move the `log.Fatal()` to `InitConfig()` to make this more explicit

romanspb80 replied:
Thanks Gilberto

Done

romanspb80 committed:
Add validation of agent flag values for ConfigMap

If the Cilium agent flags are passed via a mounted ConfigMap
(cilium-agent --config-dir=/tmp/cilium/config-map), the default for Helm
deployments, the flag values are not validated. For example if you set
"restore" with invalid value "0SO##ME5_RANDOM" in ConfigMap then Agent
would run with incorrect parameter:
.....
level=info msg=" --restore='0SO##ME5_RANDOM'" subsys=daemon
.....

But if start Agent with CLI then the validation will warn and prevent
starting the agent:
cilium-agent[8654]: invalid argument "0SO##ME5_RANDOM" for "--restore"
flag: strconv.ParseBool: parsing "0SO##ME5_RANDOM": invalid syntax

This commit add agent flag values validation for ConfigMap

Fixes: #13070

Signed-off-by: Roman Ptitcyn romanspb@yahoo.com

nathanjsweet commented on pkg/option/config.go:
```diff
@@ -3164,8 +3165,27 @@ func sanitizeIntParam(paramName string, paramDefault int) int {
+			}
+		}
+	}
+
```
nit: trailing whitespace.

romanspb80 replied:
> nit: trailing whitespace.

I use gofmt before commit

nathanjsweet commented on pkg/option/config.go:
```diff
@@ -3164,8 +3165,27 @@ func sanitizeIntParam(paramName string, paramDefault int) int {
+			// check whether the flag exists
+			if flag := flags.Lookup(key); flag != nil {
+				// validate the value of flag
+				if err := flag.Value.Set(val); err != nil {
```
Is this going to overwrite values that might be set in the command line? This seems unexpected to me.

romanspb80 replied:
> Is this going to overwrite values that might be set in the command line? This seems unexpected to me.

No, it isn't. I checked it. For example, in case of bool type of value will be done this function: https://github.com/spf13/pflag/blob/master/bool.go#L20
It doesn't overwrite values from CLI. Also the function "validateConfigmap"  is executed only for ConfigMap
